src/http/http_builder.h

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "http_consts.h"
+#include "tls/base64.h"
+#include "tls/hash.h"
+
+#include <fmt/format_header_only.h>
+#include <http-parser/http_parser.h>
+#include <map>
+#include <string>
+#include <vector>
+
+namespace http
+{
+  using HeaderMap = std::map<std::string, std::string, std::less<>>;
+
+  static std::string get_header_string(const HeaderMap& headers)
+  {
+    std::string header_string;
+    for (const auto& [k, v] : headers)
+    {
+      header_string += fmt::format("{}: {}\r\n", k, v);
+    }
+
+    return header_string;
+  }
+
+  static http_method http_method_from_str(const char* s)
+  {
+#define XX(num, name, string) \
+  if (strcmp(s, #string) == 0) \
+  { \
+    return http_method(num); \
+  }
+    HTTP_METHOD_MAP(XX)
+#undef XX
+
+    throw std::logic_error(fmt::format("Unknown HTTP method '{}'", s));
+  }
+
+  class Message
+  {
+  protected:
+    HeaderMap headers;
+    const std::vector<uint8_t>* body = nullptr;
+
+    Message() = default;
+
+  public:
+    HeaderMap get_headers() const
+    {
+      return headers;
+    }
+
+    void set_header(std::string k, const std::string& v)
+    {
+      // Store all headers lower-cased to simplify case-insensitive lookup
+      std::transform(k.begin(), k.end(), k.begin(), [](unsigned char c) {
+        return std::tolower(c);
+      });
+      headers[k] = v;
+    }
+
+    void clear_headers()
+    {
+      headers.clear();
+    }
+
+    void set_body(const std::vector<uint8_t>* b)
+    {
+      using namespace http::headers;
+
+      body = b;
+
+      headers[CONTENT_LENGTH] = fmt::format("{}", body->size());
+
+      tls::HashBytes body_digest;
```

Comment:
Why are we doing this for every request?

Reply:
Good point - we only need this when signing, so I'll move it to be created when signing.